src/function_classes/retrieval.py

```diff
@@ -1,22 +1,31 @@
 import torch
 import torch.distributions as D
 
+from torch import Tensor
+
 from core import FunctionClass
+from utils import CombinedDistribution
 
 class Retrieval(FunctionClass):
 
     def _init_param_dist(self) -> D.Distribution:
-        return D.Categorical(torch.Tensor([1]))
-    
-    def evaluate(self, x_batch: torch.Tensor, *params: torch.Tensor) -> torch.Tensor:
-        assert x_batch.shape[0] == self.batch_size and x_batch.shape[2] == self.x_dim
-        assert x_batch.shape[1] % 2 == 1
-
-        keys_b, values_b = x_batch[:, :-1:2, :], x_batch[:, 1::2, :]
-        query_b = x_batch[:, -1, :].unsqueeze(dim=1)
-        inner_products = torch.bmm(query_b, keys_b.transpose(1, 2)).squeeze(1)
-        _, retrieval_inds = torch.max(inner_products, dim=1)
-        retrieval_inds = retrieval_inds.view(-1, 1, 1).expand(-1, -1, values_b.size(-1))
-        y_batch = torch.gather(values_b, 1, retrieval_inds).squeeze(1)
-        
-        return y_batch
+        return CombinedDistribution(
+            self.x_dist,
+            D.Categorical(torch.ones([self.x_dim]) / self.x_dim)
```

Comment:
wait, don't we want to uniformly get a query_idx from one of the keys in the sequence? so for `seq_len=N`, we want to uniformly pick at random from $1,2,\ldots,N$?

Reply:
as opposed to picking an index within one of the x in the xs (which would correspond to picking a part of a key, right?)

Reply:
Op, good catch